stl/inc/memory

```diff
@@ -4094,6 +4094,167 @@ public:
 
+#if _HAS_CXX23 && defined(__cpp_lib_concepts)
+template <class _Ty>
+struct _Pointer_of_helper {};
+
+// clang-format off
+template <class _Ty>
+    requires requires { typename _Ty::pointer; }
+struct _Pointer_of_helper<_Ty> {
+    using type = typename _Ty::pointer;
+};
+
+template <class _Ty>
+    requires (!requires { typename _Ty::pointer; } && requires { typename _Ty::element_type; })
+struct _Pointer_of_helper<_Ty> {
+    using type = typename _Ty::element_type*;
+};
+
+template <class _Ty>
+    requires (!requires { typename _Ty::element_type; } && !requires { typename _Ty::pointer; }
+        && requires { typename pointer_traits<_Ty>::element_type; })
+struct _Pointer_of_helper<_Ty> {
+    using type = typename pointer_traits<_Ty>::element_type*;
+};
+// clang-format on
+
+template <class _Ty>
+using _Pointer_of = typename _Pointer_of_helper<_Ty>::type;
+
+template <class _Ty, class _Uty>
+struct _Pointer_of_or_helper {
+    using type = _Uty;
+};
+
+// clang-format off
+template <class _Ty, class _Uty>
+    requires requires { typename _Pointer_of<_Ty>; }
+struct _Pointer_of_or_helper<_Ty, _Uty> {
+    using type = _Pointer_of<_Ty>;
+};
+// clang-format on
+
+template <class _Ty, class _Uty>
+using _Pointer_of_or = typename _Pointer_of_or_helper<_Ty, _Uty>::type;
+
+template <class _SmartPtr, class _Sp, class _Pointer, class... _ArgsT>
+concept _Resettable_pointer = requires(_SmartPtr& _Smart_ptr, _Pointer _Ptr, _ArgsT&&... _Args) {
+    _Smart_ptr.reset(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args)...);
+};
+
+template <class _SmartPtr, class _Pointer, class... _ArgsT>
+class out_ptr_t {
+    static_assert(!_Is_specialization_v<_SmartPtr, shared_ptr> || sizeof...(_ArgsT) != 0,
+        "out_ptr_t with shared_ptr requires a deleter");
+
+public:
+    explicit out_ptr_t(_SmartPtr& _Smart_ptr_, _ArgsT... _Args_)
+        : _Smart_ptr(_Smart_ptr_), _Args(_STD forward<_ArgsT>(_Args_)...) {}
+
+    out_ptr_t(const out_ptr_t&) = delete;
+
+    ~out_ptr_t() {
+        if (_Ptr) {
+            _STD apply(
+                [&](auto&&... _Args_) {
+                    using _Sp = _Pointer_of_or<_SmartPtr, _Pointer>;
+                    if constexpr (_Resettable_pointer<_SmartPtr, _Sp, _Pointer, _ArgsT...>) {
+                        _Smart_ptr.reset(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args_)...);
+                    } else {
+                        _Smart_ptr = _SmartPtr(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args_)...);
+                    }
+                },
+                _STD move(_Args));
+        }
+    }
+
+    operator _Pointer*() const noexcept {
+        return _STD addressof(const_cast<_Pointer&>(_Ptr));
+    }
+
+    operator void**() const noexcept requires(!is_same_v<_Pointer, void*>) {
+        static_assert(is_pointer_v<_Pointer>, "out_ptr_t's operator void** requires Pointer to be a pointer");
+        return reinterpret_cast<void**>(_STD addressof(const_cast<_Pointer&>(_Ptr)));

```

Comment:
We have the precondition that we should not have called `operator _Pointer*` Do we care enough to actually validate this somehow? It would most likely cost some memory so I am totally fine with nasal demons emerging

Reply:
I don't think it's worth validating. It will be used as `foo(out_ptr(my_ptr))` 99.9% of the time.

Reply:
I agree with Adam. It's not so much the memory as it is having the ABI of `out_ptr` differ between debug and release modes. We already do that with many types, but doing it here will probably cause more harm than good, given Adam's point about common usage